### DISCUSSION TITLE
Help with DXF Analysis

### DISCUSSION THREAD

**Discussion by oliexe** [mozman/ezdxf#391] [Q&A]
Hi,

I am totally new to this stuff but I am searching for a way to calculate (approximately) the following attributes of a sheet metal part (a relatively simple one) in the DXF file :

+ length & weight
+ part area
+ part weight (I know the material thickness and weight)
+ number of pierces

I was trying to figure it on my own, doing numerous searches around the internet but cant seem to get started and get some solid results. Can somebody point me into the right direction ?

Thank You.

**Comment by mbway** [comment]
could you give some more specific details, like an example file you want to work with?
- does the dxf file use real world units or does it use arbitrary units and a scale bar embedded in the file?
- is there something distinctive you can use to identify the parts? like maybe they all exist on a specific layer?
- what have you tried already? do you have any code so far?

**Reply by oliexe** [reply]
Hi,

- DXF is using real-world units
- The whole sheet metal part is on one

I have followed the following resources:
https://stackoverflow.com/questions/48919164/how-to-calculate-a-dxfs-used-area-in-python-3-with-ezdxf-or-dxfgrabber
Trying to work with EXTMIN and EXTMAX values. 

But I have found that box is available and was curious that maybe that can be a good way to go:
https://ezdxf.mozman.at/docs/bbox.html

**Reply by mbway** [reply]
EXTMIN and EXTMAX are basically useless, I'm not sure what their intended purpose is but they rarely correspond to the contents of the file itself.

getting the bounding box should be fine if you know that
1. the file only contains the object you want to measure
2. the object is always axis-aligned

if either of these don't hold then you can try to [identify the entities](https://ezdxf.mozman.at/docs/tutorials/getting_data.html) which correspond to your object and calculate it's area with those entities.

**Reply by oliexe** [reply]
> EXTMIN and EXTMAX are basically useless, I'm not sure what their intended purpose is but they rarely correspond to the contents of the file itself.
> 
> getting the bounding box should be fine if you know that
> 
> 1. the file only contains the object you want to measure
> 2. the object is always axis-aligned
> 
> if either of these don't hold then you can try to [identify the entities](https://ezdxf.mozman.at/docs/tutorials/getting_data.html) which correspond to your object and calculate it's area with those entities.

Well, the rule for the script should be that the input file contains ONLY the object (+ correctly aligned) I want to measure.

**Reply by Timerlen** [reply]
Give me a high five on this, too. BBox is brilliant. Somewhere in the previous posts there will be a reply from the creator. Great respect for him! I think about the number of raises, I already had it and it even counted nicely, but bad luck and deleted the file. I tried on SPLINE because it was the easiest one, but the problem appeared on open objects and on those marked with color. The color includes the elements for marking. As you only have a cut, the replacement of lines, circles and the rest on SPLINE is ok. Also in previous posts you can find my question

**Reply by oliexe** [reply]
@mbway 
So I successfully run some tests with bounding boxes. Works nice for size estimation :)

Are you aware of any technique that can produce area and weight estimation (I know the part thickness and material weight)? Bounding boxes produces nice rough results, but for example, if I have a simple cog it still has a lot of spare material left (since it's just a box). Any idea how can I improve the results a bit more so I can calculate better estimate?

**Reply by mbway** [reply]
I thought it went without saying that bounding boxes are only suitable for axis-aligned rectangular shapes. Since you mentioned them I thought you were only dealing with rectangles. if you want to handle more complex geometry then you will have to process the geometry, not just approximate the bounding box.
how best to do that really depends on what entities you are dealing with. If you are dealing with polygons then the solution will have to be very different to if you were dealing with outlines.

I would suggest you have a look at what files you have to deal with and figure out a way to extract the geometry information of the shapes you want to measure. I would suggest looking at the recently added [disassemble module](https://ezdxf.readthedocs.io/en/stable/disassemble.html) for this.

If the shapes are filled and you really don't want to iterate over the entities then I have a more hacky solution that should work for all types of geometry: render the dxf file to an image, recording the CAD to pixels transformation (specifically the scale). the calculate the area in pixels and convert that to real world units.
